snmp/datadog_checks/snmp/snmp.py

```diff
@@ -321,7 +333,9 @@ def _start_discovery(self):
             message = 'discovery_interval could not be parsed as a number: {!r}'.format(raw_discovery_interval)
             raise ConfigurationError(message)
 
-        self._thread = threading.Thread(target=self.discover_instances, args=(discovery_interval,), name=self.name)
+        self._thread = self._thread_factory(
+            target=discover_instances, args=(self._config, discovery_interval, weakref.ref(self)), name=self.name
```

Comment:
IMO we need a big fat comment to explain why we're using a `weakref` to the check here (it's very odd-looking but necessary), and why this isn't a check method anymore.

Or this can also be explained in a docstring for the `discover_instances()` function (probably better), where we explain what the `check_ref` is and why we need one.

Reply:
Done I think.

Reply:
Yes, looks great!